### PR TITLE
Generate thunks for address-taken imports

### PR DESCRIPTION
**Pull request by dschuff**
Under emscripten, C code can take the address of a function implemented
in Javascript (which is exposed via an import in wasm). Because imports
do not have linear memory address in wasm, we need to generate a thunk
to be the target of the indirect call; it call the import directly.

This is facilited by a new .s directive (.functype) which declares the
types of functions which are declared but not defined.

Fixes https://github.com/WebAssembly/binaryen/issues/392

### CODE REVIEW

File: src/s2wasm.h [hunk 1]
```diff
@@ -480,6 +486,21 @@ class S2WasmBuilder {
     skipWhitespace();
   }
 
+  void parseFuncType() {
+    FunctionType* decl = new FunctionType;//allocator->alloc<FunctionType>();
```
Comment (kripken) [inline src/s2wasm.h]: unnecessary comment

File: src/s2wasm.h [hunk 2]
```diff
@@ -480,6 +486,21 @@ class S2WasmBuilder {
...
+    decl->name = getCommaSeparated();
+    skipComma();
+    if(match("void")) {
```
Comment (kripken) [inline src/s2wasm.h]: space after `if`

File: src/wasm.h [hunk 3]
```diff
@@ -1428,6 +1428,7 @@ class Module {
   std::vector<std::unique_ptr<Import>> imports;
   std::vector<std::unique_ptr<Export>> exports;
   std::vector<std::unique_ptr<Function>> functions;
+  std::vector<std::unique_ptr<FunctionType>> externTypes;
```
Comment (kripken) [inline src/wasm.h]: What is an "extern type"? The 4 things above it are actual properties of modules. Is this something for internal use? It seems like it should have a different API and not be alongside those 4. But, I'm not sure yet I even see the need for it - why not just use a function type?

Comment (dschuff) [inline src/wasm.h]: externTypes are the types of functions which are declared but not defined. You're right that it's simpler to keep the types along with the other function types, I fixed that. `externTypesMap` now just maps the declaration name to the type.

File: src/s2wasm.h (hunk 1 above)
Comment (dschuff) [inline src/s2wasm.h]: Fixed.

File: src/s2wasm.h (hunk 2 above)
Comment (dschuff) [inline src/s2wasm.h]: Fixed.

File: src/wasm.h [hunk 4]
```diff
@@ -1441,6 +1441,8 @@ class Module {
   std::map<Name, Import*> importsMap;
   std::map<Name, Export*> exportsMap;
   std::map<Name, Function*> functionsMap;
+  // Types of functions which are declared but not defined.
+  std::map<Name, FunctionType*> externTypesMap;
```
Comment (kripken) [inline src/wasm.h]: I don't think we even need this? This is an implementation detail of llvm/s2wasm, it shouldn't cause additional complexity in wasm.h for all modules generated in all methods. That is, can't llvm/s2wasm just emit a normal function type for this, and use it like a normal function type - why do we need this map?

Comment (dschuff) [inline src/wasm.h]: It does need to exist somewhere; a functionType is just a type (it can have a name, but that' the name for the type, i.e. the asm.js-style signature). We still need to have a list of the external functions that are not part of the module but that the linker knows how to generate thunks for. You're right that it doesn't necessarily need to be on the module. It could be in the linker instead. Although in that case it would be weird to have the types in the module, and the functions in the linker; probably would put everything in the linker.

Comment (kripken) [inline src/wasm.h]: Hmm, I think it should all be in the linker, then.

Comment (dschuff) [inline src/wasm.h]: OK, I moved the mapping to the linker but decided to let the wasm module continue to own the function type itself, to allow for deduping and because it already has the unique_ptrs.

File: src/wasm-linker.cpp [hunk 5]
```diff
@@ -362,9 +374,26 @@ void Linker::makeDynCallThunks() {
...
     f->body = call;
     out.wasm.addFunction(f);
     exportFunction(f->name, true);
   }
 }
+
+Function* Linker::generateImportThunk(Name name, const FunctionType* funcType) {
+  wasm::Builder wasmBuilder(out.wasm);
+  std::vector<NameType> params;
+  int p = 0;
```
Comment (kripken) [inline src/wasm-linker.cpp]: could be `Index` (could matter in the future, when wasm64 has 64-bit Index, and the person is building on a 32-bit machine)

File: src/wasm-linker.cpp [hunk 6]
```diff
@@ -362,9 +374,26 @@ void Linker::makeDynCallThunks() {
...
+  for (const auto& ty : funcType->params) params.emplace_back(std::to_string(p++), ty);
+  Function *f = wasmBuilder.makeFunction(std::string("__importThunk_") + name.c_str(), std::move(params), funcType->result, {});
+  std::vector<Expression*> args;
+  for (unsigned i = 0; i < funcType->params.size(); ++i) {
```
Comment (kripken) [inline src/wasm-linker.cpp]: ditto could be `Index`

File: src/asm_v_wasm.h [hunk 7]
```diff
@@ -29,12 +29,13 @@ AsmType wasmToAsmType(WasmType type);
...
 
-std::string getSig(FunctionType *type);
+std::string getSig(const FunctionType *type);
 
 std::string getSig(Function *func);
 
-template<typename CallBase>
-std::string getSig(CallBase *call) {
+template<typename T,
+         typename std::enable_if<std::is_base_of<Expression, T>::value>::type* = nullptr>
```
Comment (kripken) [inline src/asm_v_wasm.h]: hmm, what's going on here? why is this needed now?

**Comment by kripken** [comment]
overall lgtm

File: src/wasm-linker.cpp (hunk 5 above)
Comment (dschuff) [inline src/wasm-linker.cpp]: ... assuming we ever support more than 2^32 locals :)

File: src/asm_v_wasm.h (hunk 7 above)
Comment (dschuff) [inline src/asm_v_wasm.h]: It seems like it was always ambiguous with `getSig(FunctionType*)` (i.e. this template could have been instantiated with `CallBase` as `FunctionType` which isn't what we wanted). But it wasn't until this PR that we started seeing errors (i.e. everywhere we pass `FunctionType*` this template is getting instantiated instead) of the function. I'm guess it's because I changed `FunctionType*` to `const FunctionType *` in the other declaration. In any case this makes a bit more explicit the intention that it only work on calls. Unfortunately there is no `CallBase` class, just Expression with 2 kinds of calls, but I think it's still an improvement (even aside from now being necessary).

File: src/wasm-linker.cpp (hunk 6 above)
Comment (dschuff) [inline src/wasm-linker.cpp]: Done.

File: src/asm_v_wasm.h (hunk 7 above)
Comment (kripken) [inline src/asm_v_wasm.h]: Ok, cool.

lgtm
